code/game/machinery/teleporter.dm

```diff
@@ -87,86 +89,70 @@
 		data["teleporterhub"] = null
 		data["calibrated"] = null
 		data["accurate"] = null
-	data["regime"] = regime_set
+	data["isGate"] = isGate
 	var/area/targetarea = get_area(target)
 	data["target"] = (!target || !targetarea) ? "None" : sanitize(targetarea.name)
 	data["calibrating"] = calibrating
 	data["locked"] = locked
 	return data
 
-/obj/machinery/computer/teleporter/Topic(href, href_list)
+/obj/machinery/computer/teleporter/tgui_act(action, params)
 	if(..())
-		return 1
-
-	if(href_list["eject"])
-		eject()
-		SSnanoui.update_uis(src)
-		return
+		return TRUE
 
 	if(!check_hub_connection())
 		to_chat(usr, "<span class='warning'>Error: Unable to detect hub.</span>")
-		SSnanoui.update_uis(src)
 		return
 	if(calibrating)
 		to_chat(usr, "<span class='warning'>Error: Calibration in progress. Stand by.</span>")
-		SSnanoui.update_uis(src)
 		return
 
-	if(href_list["regimeset"])
-		power_station.engaged = 0
-		power_station.teleporter_hub.update_icon()
-		power_station.teleporter_hub.calibrated = 0
-		reset_regime()
-		SSnanoui.update_uis(src)
-	if(href_list["settarget"])
-		power_station.engaged = 0
-		power_station.teleporter_hub.update_icon()
-		power_station.teleporter_hub.calibrated = 0
-		set_target(usr)
-		SSnanoui.update_uis(src)
-	if(href_list["lock"])
-		power_station.engaged = 0
-		power_station.teleporter_hub.update_icon()
-		power_station.teleporter_hub.calibrated = 0
-		target = get_turf(locked.locked_location)
-		SSnanoui.update_uis(src)
-	if(href_list["calibrate"])
-		if(!target)
-			to_chat(usr, "<span class='warning'>Error: No target set to calibrate to.</span>")
-			SSnanoui.update_uis(src)
-			return
-		if(power_station.teleporter_hub.calibrated || power_station.teleporter_hub.accurate >= 3)
-			to_chat(usr, "<span class='notice'>Hub is already calibrated.</span>")
-			SSnanoui.update_uis(src)
+	switch(action)
+		if("eject")
+			eject()
 			return
-		src.visible_message("<span class='notice'>Processing hub calibration to target...</span>")
-
-		calibrating = 1
-		SSnanoui.update_uis(src)
-		spawn(50 * (3 - power_station.teleporter_hub.accurate)) //Better parts mean faster calibration
-			calibrating = 0
-			if(check_hub_connection())
-				power_station.teleporter_hub.calibrated = 1
-				src.visible_message("<span class='notice'>Calibration complete.</span>")
-			else
-				src.visible_message("<span class='warning'>Error: Unable to detect hub.</span>")
-			SSnanoui.update_uis(src)
+		if("regimeset")
+			power_station.engaged = FALSE
+			power_station.teleporter_hub.calibrated = FALSE
+			target = null
+			isGate = !isGate //switches between teleporter and gate
+			power_station.teleporter_hub.update_icon()
+		if("settarget")
+			power_station.engaged = FALSE
+			power_station.teleporter_hub.update_icon()
+			power_station.teleporter_hub.calibrated = FALSE
```

Comment:
This block of code seems to be reoccurring at least 3 times. Please try to eliminate all duplicate code by making new procs if needed